File: agent-driven-architecture-main/examples/typescript/src/spine/pure/turn.ts
// ── spine/pure/turn — how a turn ENDS, and what the consumer must report ────
// ZERO I/O. Two sealed sets, both consumed with `never`-guarded matches.
//
// `TurnOutcome` is 12.4 made typed: a turn that throws does not kill the
// consumer and does not vanish. It DEGRADES to a status carrying its cause, the
// consumer folds that cause, and the next message is processed. The consumer is
// the heartbeat and it does not stop.
//
// `ConsumerEvent` is what the consumer must SAY OUT LOUD. Every one of these is
// a decision the system took on the operator's behalf — a dropped input, a
// duplicate refused, a turn that failed, a cancel that timed out — and 12.2's
// "observable, never silent" is only true if each one travels the ordinary path:
// resolveAction → gate → fold → commit → signed Command. Reporting is therefore
// a REQUIRED consumer parameter, which is default-deny applied to observability.

import type { SourceKey, SourceName } from "./ids";

// ── How a turn ended ────────────────────────────────────────────────────────

export type TurnOutcomeKind = "Ok" | "Threw" | "Cancelled" | "Idle";

export interface TurnOutcomeBase {
  readonly kind: TurnOutcomeKind;
}

export interface TurnOk extends TurnOutcomeBase {
  readonly kind: "Ok";
  readonly steps: number;
}

/** The turn threw. The CONSUMER LIVES (12.4) — the exception never crosses the
 *  loop, because the turn runner captures it into this variant. */
export interface TurnThrew extends TurnOutcomeBase {
  readonly kind: "Threw";
  readonly fault: string;
}

/** Preempted at a step boundary. Steps that completed before the cancel stay
 *  durably folded and their effects stay performed — there is no rollback. */
export interface TurnCancelled extends TurnOutcomeBase {
  readonly kind: "Cancelled";
  readonly by: SourceName;
}

/** "Nothing to do" is not a failure. */
export interface TurnIdle extends TurnOutcomeBase {
  readonly kind: "Idle";
}

export type TurnOutcome = TurnOk | TurnThrew | TurnCancelled | TurnIdle;

export function turnOk(steps: number): TurnOk {
  return { kind: "Ok", steps };
}

export function turnThrew(fault: string): TurnThrew {
  return { kind: "Threw", fault };
}

export function turnCancelled(by: SourceName): TurnCancelled {
  return { kind: "Cancelled", by };
}

export const turnIdle: TurnIdle = { kind: "Idle" };

// ── What the consumer must never swallow ────────────────────────────────────

export type ConsumerEventKind = "Conflated" | "Duplicate" | "TurnFailed" | "CancelDeadlineExceeded";

export interface ConsumerEventBase {
  readonly kind: ConsumerEventKind;
  readonly source: SourceName;
}

/** A busy-drop on a perishable source. `dropped` is the count folded into the
 *  turn that finally ran — so the reasoner is told it is shedding load. */
export interface Conflated extends ConsumerEventBase {
  readonly kind: "Conflated";
  readonly dropped: number;
}

/** A durable source re-delivered an item already committed. */
export interface Duplicate extends ConsumerEventBase {
  readonly kind: "Duplicate";
  readonly key: SourceKey;
}

export interface TurnFailed extends ConsumerEventBase {
  readonly kind: "TurnFailed";
  readonly fault: string;
}

/** The cancelled turn did not settle inside its bound, so it was ABANDONED:
 *  its channel into the system was revoked and the new turn started anyway.
 *  Named, degraded and counted — never hidden. */
export interface CancelDeadlineExceeded extends ConsumerEventBase {
  readonly kind: "CancelDeadlineExceeded";
  readonly afterMs: number;
}

export type ConsumerEvent = Conflated | Duplicate | TurnFailed | CancelDeadlineExceeded;

export function conflated(source: SourceName, dropped: number): Conflated {
  return { kind: "Conflated", source, dropped };
}

export function duplicate(source: SourceName, key: SourceKey): Duplicate {
  return { kind: "Duplicate", source, key };
}

export function turnFailed(source: SourceName, fault: string): TurnFailed {
  return { kind: "TurnFailed", source, fault };
}

export function cancelDeadlineExceeded(
  source: SourceName,
  afterMs: number,
): CancelDeadlineExceeded {
  return { kind: "CancelDeadlineExceeded", source, afterMs };
}

File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/pure/Turn.kt
// ── spine/pure/turn — how a turn ENDS, and what the consumer must not swallow ─
// 12.4: a turn that throws degrades to a TYPED STATUS carrying its cause, and the
// consumer lives. The exception never crosses the loop, because the loop is the
// heartbeat and a heartbeat that dies on a bad message is not a heartbeat.
//
// ConsumerEvent is the other half of the same rule: a busy-drop, a duplicate, a
// fault and a blown cancel deadline are all DECISIONS, so none of them may be
// silent (12.2). They are reported out of the consumer as spine-shaped events; the
// composition root maps them to app Actions, and they then travel the ONE existing
// path — resolveAction → gate → fold → commit → signed Command. That is why
// observability here costs the core zero new machinery.
//
// ZERO I/O (gate check C8).

package adr.spine.pure

/** How one turn ended. Closed, so every consumer of an outcome is compiler-checked. */
sealed class TurnOutcome {
    /** It ran to completion, submitting [steps] steps through the boundary. */
    data class Ok(val steps: Int) : TurnOutcome()

    /** 12.4: the turn threw. The cause is carried; THE CONSUMER LIVES. */
    data class Threw(val fault: String) : TurnOutcome()

    /** Preempted at a step boundary. Steps completed before the cancel STAY folded. */
    data class Cancelled(val by: SourceName) : TurnOutcome()

    /** Nothing to do. "Idle" is a status, not a failure. */
    data object Idle : TurnOutcome()
}

/**
 * What the consumer must report rather than swallow. Each one folds as a signed
 * Command through the root's mapping, so "observable, never silent" is a property
 * of the timeline rather than of a log line someone might grep for.
 */
sealed class ConsumerEvent(
    /** Which source the event is about. Declared once (G12). */
    open val source: SourceName,
) {

    /** A perishable busy-drop: [dropped] inputs were conflated away (12.2). */
    data class Conflated(override val source: SourceName, val dropped: Int) : ConsumerEvent(source)

    /** A durable redelivery whose key had already been folded — deduped, then acked. */
    data class Duplicate(override val source: SourceName, val key: SourceKey) : ConsumerEvent(source)

    /** 12.4: a turn threw, and this is its cause. */
    data class TurnFailed(override val source: SourceName, val fault: String) : ConsumerEvent(source)

    /**
     * The cancel deadline blew: the turn ignored cancellation, so its submit channel
     * was REVOKED and it was abandoned. The leak is named, degraded and counted —
     * removing it would need an unbounded join, which 12.3 itself calls a hang.
     */
    data class CancelDeadlineExceeded(
        override val source: SourceName,
        val afterMs: Millis,
    ) : ConsumerEvent(source)
}
